src/diff/index.js

```diff
@@ -554,14 +554,26 @@ function diffElementNodes(
 
 /**
  * Invoke or update a ref, depending on whether it is a function or object ref.
- * @param {Ref<any>} ref
+ * @param {Ref<any> & { _unmount?: unknown }} ref
  * @param {any} value
  * @param {VNode} vnode
  */
 export function applyRef(ref, value, vnode) {
 	try {
-		if (typeof ref == 'function') ref(value);
-		else ref.current = value;
+		if (typeof ref == 'function') {
+			let hasRefUnmount = typeof ref._unmount == 'function';
+			if (hasRefUnmount) {
+				// @ts-ignore TS doesn't like moving narrowing checks into variables
+				ref._unmount();
+			}
+
+			if (!hasRefUnmount || value != null) {
```

Comment:
Why do we need `!hasRefUnmount` here? Wouldn't it be safe to always invoke this and inline `typeof ref._unmount == 'function'` above?

Reply:
Maybe inverting the check makes it more clear. When a cleanup function is returned, then the `ref` function should never be invoked with `null`. But when no cleanup function is there it should be invoked with `null`.